src/ray/core_worker/task_manager.cc

```diff
@@ -1087,9 +1087,22 @@ bool TaskManager::RetryTaskIfPossible(const TaskID &task_id,
         RAY_CHECK(num_oom_retries_left == 0);
       }
     } else {
-      if (num_retries_left > 0) {
+      auto is_preempted = false;
+      if (error_info.error_type() == rpc::ErrorType::NODE_DIED) {
+        const auto node_info = gcs_client_->Nodes().Get(task_entry.GetNodeId(),
+                                                        /*filter_dead_nodes=*/false);
+        is_preempted = node_info != nullptr && node_info->has_death_info() &&
+                       node_info->death_info().reason() ==
+                           rpc::NodeDeathInfo::AUTOSCALER_DRAIN_PREEMPTED;
+      }
+      if (num_retries_left > 0 || (is_preempted && task_entry.spec.IsRetriable())) {
```

Comment:
Even when num_retries_left == 0, we should still retry the task if it was preempted.